experiment/secret-sync-controller/main_test.go

```diff
@@ -1,295 +1,435 @@
+    secret:
+    - name: secret-a
+  - name: ns-c
 
-func (cl clientMock) UpsertKubernetesSecret(spec KubernetesSpec, data *SecretData) (*SecretData, error) {
-	_, ok := cl.K8sSecret[spec.Namespace][spec.Secret]
-	if !ok {
-		return nil, fmt.Errorf("Secret not found: K8s:/namespaces/%s/secrets/%s", spec.Namespace, spec.Secret)
-	}
-	cl.K8sSecret[spec.Namespace][spec.Secret][spec.Key] = data.Data
-	return data, nil
+*/
+
+var setup = client.Setup{
+	SecretManager: client.SecretManagerSetup{
+		Projects: []client.ProjectSetup{
+			{
+				Name: "k8s-jkns-gke-soak",
+				Secrets: []client.GSMSecretSetup{
+					{
+						Name: "gsm-password",
+						Data: "gsm-password-v1",
+					},
+					{
+						Name: "gsm-token",
+						Data: "gsm-token-v1",
+					},
+				},
+			},
+		},
+	},
+	Kubernetes: client.KubernetesSetup{
+		Namespaces: []client.NamespaceSetup{
+			{
+				Name: "ns-a",
+				Secrets: []client.K8sSecretSetup{
+					{
+						Name: "secret-a",
+						Keys: []client.KeySetup{
+							{
+								Name: "key-a",
+								Data: "old-token",
+							},
+						},
+					},
+				},
+			},
+			{
+				Name: "ns-b",
+				Secrets: []client.K8sSecretSetup{
+					{
+						Name: "secret-b",
+					},
+				},
+			},
+			{
+				Name: "ns-c",
+			},
+		},
+	},
 }
 
 func TestSync(t *testing.T) {
 
 	var tests = []struct {
-		input     SecretSyncSpec
-		want      []byte
+		config    SecretSyncConfig
+		want      client.KubernetesSetup
 		expectErr bool
 	}{
 		{
-			input: SecretSyncSpec{
-				Source: SecretManagerSpec{
-					Project: "projA",
-					Secret:  "secretA",
+			// happy path
+			// syncing from <existing gsm secret> to <existing k8s secret> with <existing key>.
+			// should update the key-value pair
+			config: SecretSyncConfig{
+				Specs: []SecretSyncSpec{
+					{
+						Source: SecretManagerSpec{
+							Project: "k8s-jkns-gke-soak",
+							Secret:  "gsm-password",
+						},
+						Destination: KubernetesSpec{
+							Namespace: "ns-a",
+							Secret:    "secret-a",
+							Key:       "key-a",
+						},
+					},
 				},
-				Destination: KubernetesSpec{
-					Namespace: "ns1",
-					Secret:    "secret1",
-					Key:       "key1",
+			},
+
+			want: client.KubernetesSetup{
+				Namespaces: []client.NamespaceSetup{
+					{
+						Name: "ns-a",
+						Secrets: []client.K8sSecretSetup{
+							{
+								Name: "secret-a",
+								Keys: []client.KeySetup{
+									{
+										Name: "key-a",
+										Data: "gsm-password-v1",
+									},
+								},
+							},
+						},
+					},
+					{
+						Name: "ns-b",
+						Secrets: []client.K8sSecretSetup{
+							{
+								Name: "secret-b",
+							},
+						},
+					},
+					{
+						Name: "ns-c",
+					},
 				},
 			},
-			want:      []byte("valueOfSecretA"),
 			expectErr: false,
 		},
 		{
-			input: SecretSyncSpec{
-				Source: SecretManagerSpec{
-					Project: "projB",
-					Secret:  "secretB",
+			// syncing from <existing gsm secret> to <existing k8s secret> with <non-existing key>.
```

Comment:
So I would be curious to see what happens if we were to sync to `secret-a`, would we see both `key-a` and `missed` in its data?  At present this could be seen as overwriting `{}` with `missed: gsm-token-v1`

Reply:
Yes we would see both keys under `secret-a`.
I'll add that test case.